src/schedule/edf_schedule.c

```diff
@@ -124,12 +129,16 @@ static void schedule_edf_task(void *data, struct task *task, uint64_t start,
 
 	/* add task to the list */
 	list_item_append(&task->list, &edf_sch->list);
+	tracev_edf_sch("schedule_edf_task() %p", (long)task);
 
 	task->state = SOF_TASK_STATE_QUEUED;
 
 	irq_local_enable(flags);
 
-	schedule_edf(data);
+	if (cpu_get_id() == PLATFORM_MASTER_CORE_ID ||
```

Comment:
@lyakh Nice catch. It's because master core doesn't wait for slave core to actually boot up before responding to power up ipc. I wonder whether more appropriate solution would be to wait for some kind of status reported by slave core after starting. This feels a little bit hacky.

Reply:
Lets bring this into a new PR as it may get bogged down here with other patches.

Reply:
@tlauda no, don't think it's because the master core doesn't wait.It doesn't matter what the master core does. The slave core while booting gets an IDC interrupt and schedules a task, before completing its own initialisation, where in the *end* of that initialisation its idle task is started.

Reply:
@lyakh The only source of the IDC is master core. During my testing I've never seen this kind of race scenario, but I guess python is too slow to send another IPC before slave core boots.

Reply:
@tlauda but it doesn't seem to be a real IDC, because with this fix I'm ignoring it and it isn't causing any problem

Reply:
@lyakh The task is still queued, just executed later.

Reply:
@tlauda I did in fact try that: I added a 50ms delay in the kernel after powering up the core before sending the IPC to inform the firmware about that. Didn't help.

Reply:
@lyakh Weird and very suspicious. So we are getting some kind of random IDC interrupt and don't even know what is the source...

Reply:
@tlauda ok, maybe that was wrong. I double-checked with the delay in the kernel between booting the slave core and sending the IPC, and the result is the same - it doesn't help. But this time I also printed in the firmware. It looks like the slave core doesn't actually boot until the IPC. Or at least it's hanging somewhere early after the power on but before initialising the EDF scheduler and only after the IDC interrupt it continues? Any idea?

Reply:
@lyakh What is the exact flow from the kernel perspective. You are powering up core, unstalling and unresetting and then send power up IPC? And after that IPC if you wait a little bit the slave core still hasn't booted?

Reply:
@tlauda as far as I can see, a slave core is enabled (SPA is set), then reset is cleared, then the core is unstalled. Then I insert a delay after which the kernel sends an IPC to the master core, but the slave core only boots after this last IPC.

Reply:
@lyakh You need to wait after first IPC and before the next one. First IPC is to actually boot slave core out of ROM and jump to the FW.

Reply:
@tlauda but then we have exactly the problem: the first IPC sends an IDC to the booting core, which then arrives before the idle task has been scheduled

Reply:
@lyakh But core isn't actually booting until you send SOF_IPC_PM_CORE_ENABLE. It is stuck in ROM in endless loop. After sending this particular IPC it boots up and initialized everything.

Reply:
@tlauda but it's already the first IPC that sends the first IDC: see `arch_cpu_enable_core()`.
```
	pm_runtime_get(PM_RUNTIME_DSP, id);
	pm_runtime_get(CORE_MEMORY_POW, id);
	cpu_alloc_core_context(id);
	idc_enable_interrupts(id, cpu_get_id());
	idc_send_msg(&power_up, IDC_NON_BLOCKING);
	active_cores_mask |= (1 << id);
```
So, do we take my fix or do we remove sending an IDC message from here or do we process this first IDC on the slave core without scheduling a task (effectively my approach)?

Reply:
@lyakh I don't get your point. First IPC sends first IDC, because it needs to boot up the slave core. Without it you won't even get to EDF scheduler initialization. The problem is the second IDC, which is for some reason sent after this one, but before core boots up.

Reply:
@tlauda ok. I thought it was the first IDC, that was already triggering the scheduler task. But that's even better then, then we really don't need to schedule in that first IDC interrupt. Maybe a better fix than special-casing in the scheduler would be not scheduling in the IDC handler if it's the POWER_UP message? That seems to be a better solution to me than waiting for some undefined time in the kernel without a way to (asynchronously) verify the core status?

Reply:
@lyakh Could you check with #2593 whether issue is gone?